src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -5,96 +5,341 @@ import org.apache.spark.rdd.RDD
 /**
  * Node-level optimization, such as selecting a Linear Solver
  *
- * @param sampleFraction The fraction of the RDD to use for operations
+ * @param samplePerPartition The number of items per partition to look at when optimizing nodes
  */
-class NodeOptimizationRule(sampleFraction: Double = 0.01, seed: Long = 0) extends Rule {
+class NodeOptimizationRule(samplePerPartition: Int = 3) extends Rule {
+  case class OptimizationState(
+    optimizedInstructions: Seq[Instruction],
+    unoptimizedInstructionIndexToOptimizedIndex: Int => Int,
+    registers: Map[Int, InstructionOutput],
+    numPerPartitionPerNode: Map[Int, Map[Int, Int]]
+  )
+
+  /**
+   * Get the set of instructions that we need to check for optimizations at.
+   *
+   * This is any instruction that either applies an optimizable transformer,
+   * or fits an optimizable estimator,
+   * and is not supposed to be executed on test data.
+   */
+  private def getInstructionsToOptimize(instructions: Seq[Instruction]): Set[Int] = {
+    instructions.zipWithIndex.collect {
+      case (TransformerApplyNode(tIndex, _), index)
+        if instructions(tIndex).isInstanceOf[OptimizableTransformer[_, _]] => index
+
+      case (EstimatorFitNode(estIndex, _), index)
+        if instructions(estIndex).isInstanceOf[OptimizableEstimator[_, _]] => index
+
+      case (EstimatorFitNode(estIndex, _), index)
+        if instructions(estIndex).isInstanceOf[OptimizableLabelEstimator[_, _, _]] => index
+
+    }.toSet -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions)
+  }
+
+  /**
+   * Execute the instruction at the given index, and output the new state of the optimization process
+   *
+   * @param optimizationState
+   * @param instruction
+   * @param index
+   * @return
+   */
+  private def executeInstruction(optimizationState: OptimizationState, instruction: Instruction, index: Int)
+  : OptimizationState = {
+    // Get the dependencies from the registers and run the instruction on them (caching RDD outputs)
+    val registers = optimizationState.registers
+    val deps = instruction.getDependencies.map(registers)
+    val instructionOutput = instruction.execute(deps) match {
+      case RDDOutput(rdd) => RDDOutput(rdd.cache())
```

Comment:
Yeah no problem, I already have such a hook in the autocacherule and I forgot about putting one here. will do!